File: apkEditor/src/main/java/com/mcal/apkeditor/activities/MainActivity.java
package com.mcal.apkeditor.activities;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.balsikandar.crashreporter.ui.CrashReporterActivity;
import com.mcal.apkeditor.ApkComposeService;
import com.mcal.apkeditor.BuildConfig;
import com.mcal.apkeditor.GlobalConfig;
import com.mcal.apkeditor.MenuListAdapter;
import com.mcal.apkeditor.R;
import com.mcal.apkeditor.data.Dialogs;
import com.mcal.apkeditor.dialogs.AppAgreementDialog;
import com.mcal.apkeditor.dialogs.ProcessingDialog;
import com.mcal.apkeditor.prj.ProjectListActivity;
import com.mcal.apkeditor.prj.ProjectListActivity2;
import com.mcal.apkeditor.utils.OnlineMessage;
import com.mcal.common.activities.CustomizedLangActivity;
import com.mcal.common.data.Preferences;
import com.mcal.common.utils.FileUtils;
import com.mcal.httpserver.HttpServiceManager;

import java.io.File;

/**
 * For apktool, look into:
 * brut.apktool\apktool-lib\src\main\java\brut\androlib\ApkDecoder.java:decode()
 * <p>
 * And then look into: brut.androlib.res.AndrolibResources
 */
public class MainActivity extends CustomizedLangActivity implements
        AdapterView.OnItemClickListener, ProcessingDialog.ProcessingInterface {

    // Native library
    static {
        System.loadLibrary("apkeditorpro");
    }

    // Used to show a dialog
    private OnlineMessage prompter;
    // For left side menu
    private DrawerLayout mDrawerLayout;
    private ListView mDrawerList;
    private ActionBarDrawerToggle mDrawerToggle;

    // modify a zip file
    public static native void modifyZip(String target, String source, String added,
                                 int len1, String removed, int len2, String replaced, int len3);

    // Verify certificate
    // When report error of "Cannot register", to make sure it is called
    //public static native int vc(Object ctx, int seed);

    public static boolean upgradedFromOldVersion(@NonNull Context ctx) {
        File f = new File(ctx.getFilesDir(), "work.xml");
        return !f.exists();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Full screen
        if (GlobalConfig.instance(this).isFullScreen()) {
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                    WindowManager.LayoutParams.FLAG_FULLSCREEN);
        }

        setContentView(R.layout.activity_main);
        setupToolbar(getString(R.string.app_name));

        initUI();

        // As pro has no network access, cannot get online message
        if (!BuildConfig.IS_PRO) {
            this.prompter = new OnlineMessage(this);
        }

        // Show license dialog
        if (BuildConfig.SHOW_AGREEMENT) {
            if (!AppAgreementDialog.appLicenseAccepted(this)) {
                new AppAgreementDialog(this);
            } else {
                initFileWithPermissionCheck();
            }
        } else {
            initFileWithPermissionCheck();
        }
    }

    @SuppressWarnings("ConstantConditions")
    private void setupToolbar(String title) {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(title);
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onResume() {
        if (!BuildConfig.IS_PRO) {
            prompter.showMessageDialog();
        }
        super.onResume();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    private void setupSlidingMenu() {
        // enabling action bar app icon and behaving it as toggle button
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(false);
        }

        mDrawerLayout = findViewById(R.id.drawer_layout);
        mDrawerList = findViewById(R.id.slider_list);

        MenuListAdapter adapter = new MenuListAdapter(this);
        mDrawerList.setAdapter(adapter);
        mDrawerList.setOnItemClickListener(this);

        mDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout,
                R.string.app_name, // nav drawer open - description for accessibility
                R.string.app_name // nav drawer close - description for accessibility
        );
        mDrawerLayout.addDrawerListener(mDrawerToggle);
    }

    private void initUI() {
        if (BuildConfig.PARSER_ONLY) {
            AppCompatImageView imageView = findViewById(R.id.logo);
            imageView.setImageResource(R.drawable.parser_logo);
        }

        // Left sliding menu
        setupSlidingMenu();

        // Select apk from folder
        AppCompatButton openApkBtn = this.findViewById(R.id.tv_select_apkfile);
        openApkBtn.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, FileListActivity.class);
            startActivity(intent);
        });

        // Select apk from app
        AppCompatButton openAppBtn = this.findViewById(R.id.tv_select_appfile);
        if (BuildConfig.DISPLAY_APP) {
            openAppBtn.setOnClickListener(v -> {
                Intent intent = new Intent(MainActivity.this, UserAppActivity.class);
                startActivity(intent);
            });
        } else {
            openAppBtn.setText(R.string.settings);
            openAppBtn.setOnClickListener(v -> {
                Intent intent = new Intent(MainActivity.this, SettingsActivity.class);
                startActivity(intent);
            });
        }

        // Odex Patcher
        AppCompatButton odexPatcherBtn = this.findViewById(R.id.tv_odex_patcher);
        odexPatcherBtn.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, OdexPatchActivity.class);
            startActivity(intent);
        });


        // Exit
        AppCompatButton exitButton = this.findViewById(R.id.tv_exit);
        exitButton.setOnClickListener(v -> new ProcessingDialog(MainActivity.this, MainActivity.this, -1).show());

        // Help
        // For APK Parser, use it as 'project'
        AppCompatButton helpButton = this.findViewById(R.id.tv_help);
        if (BuildConfig.PARSER_ONLY) {
            helpButton.setText(R.string.projects);
            helpButton.setOnClickListener(v -> {
                Class<?> cls = (BuildConfig.PARSER_ONLY ? ProjectListActivity2.class : ProjectListActivity.class);
                Intent helpIntent = new Intent(MainActivity.this, cls);
                startActivity(helpIntent);
            });
        } else {
            helpButton.setOnClickListener(v -> {
                Intent helpIntent = new Intent(MainActivity.this, HelpActivity.class);
                startActivity(helpIntent);
            });
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // toggle nav drawer on selecting action bar app icon/title
        if (mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        if (id == R.id.action_settings) {
            Intent i = new Intent(this, SettingsActivity.class);
            startActivity(i);
            return true;
        } else if (id == R.id.action_about) {
            //Dialogs.about(this);
            Intent intent = new Intent(MainActivity.this, AboutActivity.class);
            startActivity(intent);
            return true;
        } else if (item.getItemId() == R.id.action_night_mode) {
            if (Preferences.isNightModeEnabled()) {
                Preferences.setNightModeEnabled(false);
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                getDelegate().applyDayNight();
            } else {
                Preferences.setNightModeEnabled(true);
                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                getDelegate().applyDayNight();
            }
        }
        return super.onOptionsItemSelected(item);
    }

    /***
     * Called when invalidateOptionsMenu() is triggered
     */
    @Override
    public boolean onPrepareOptionsMenu(@NonNull Menu menu) {
        // if nav drawer is opened, hide the action items
        boolean drawerOpen = mDrawerLayout.isDrawerOpen(mDrawerList);
        menu.findItem(R.id.action_settings).setVisible(!drawerOpen);
        menu.findItem(R.id.action_about).setVisible(!drawerOpen);
        return super.onPrepareOptionsMenu(menu);
    }

    /**
     * When using the ActionBarDrawerToggle, you must call it during
     * onPostCreate() and onConfigurationChanged()...
     */

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        // Sync the toggle state after onRestoreInstanceState has occurred.
        mDrawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        // Pass any configuration change to the drawer toggls
        mDrawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public void onBackPressed() {
        this.finish();
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            initFile();
        }
    }

    public void initFileWithPermissionCheck() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            int WRITE_EXTERNAL_STORAGE_REQUEST_CODE = 1;
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    WRITE_EXTERNAL_STORAGE_REQUEST_CODE);
        } else {
            initFile();
        }
    }

    private void initFile() {
        // For free version, do not need to create these files
        if (!BuildConfig.IS_PRO) {
            return;
        }

        // If need to limit the new version, does not need to create such files
        if (BuildConfig.LIMIT_NEW_VERSION) {
            return;
        }
    }

    // Click on the menu item
    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int position, long id) {
        switch ((int) id) {
            case MenuListAdapter.ITEM_PROJECT: {
                Class<?> cls = (BuildConfig.PARSER_ONLY ? ProjectListActivity2.class : ProjectListActivity.class);
                Intent intent = new Intent(this, cls);
                startActivity(intent);
                break;
            }
            case MenuListAdapter.ITEM_SETTING: {
                Intent intent = new Intent(this, SettingsActivity.class);
                startActivity(intent);
                break;
            }
            case MenuListAdapter.ITEM_IMG_DOWNLOADER: {
                Intent intent = new Intent(this, ImageDownloadActivity.class);
                startActivity(intent);
                break;
            }
            case MenuListAdapter.ITEM_ABOUT: {
                Dialogs.about(this);
                break;
            }
            case MenuListAdapter.ITEM_LOGS: {
                Intent intent = new Intent(this, CrashReporterActivity.class);
                startActivity(intent);
                break;
            }
            case MenuListAdapter.ITEM_FORUM: {
                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://forum.timscriptov.ru/index.php?apkeditor-pro.9/")));
                break;
            }
            case MenuListAdapter.ITEM_TELEGRAM: {
                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/apkeditor2021")));
                break;
            }
        }
        mDrawerLayout.closeDrawer(mDrawerList);
    }

    // Deal with exit
    @Override
    public void process() throws Exception {
        try {
            Intent intent = new Intent(MainActivity.this, ApkComposeService.class);
            stopService(intent);
            HttpServiceManager.instance().stopWebService(this);

            File fileDir = MainActivity.this.getFilesDir();
            String rootDirectory = fileDir.getAbsolutePath();
            String decodeRootPath = rootDirectory + "/decoded";
            FileUtils.deleteAll(new File(decodeRootPath));
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    // Deal with exit
    @Override
    public void afterProcess() {
        android.os.Process.killProcess(android.os.Process.myPid());
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/activities/MainActivity.kt
package com.mcal.apkeditor.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Process
import android.util.Log
import android.view.*
import android.widget.LinearLayout
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.MenuProvider
import androidx.lifecycle.lifecycleScope
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.apkeditor.ApkComposeService
import com.mcal.apkeditor.BuildConfig
import com.mcal.apkeditor.R
import com.mcal.apkeditor.databinding.ActivityMainBinding
import com.mcal.apkeditor.dialogs.AppAgreementDialog
import com.mcal.apkeditor.dialogs.AppAgreementDialog.Companion.appLicenseAccepted
import com.mcal.apkeditor.dialogs.selectFullEditDialog
import com.mcal.apkeditor.prj.ProjectListActivity
import com.mcal.apkeditor.settings.presentation.SettingsActivity
import com.mcal.apkeditor.utils.Utils
import com.mcal.common.App
import com.mcal.common.activities.CustomizedLangActivity
import com.mcal.common.data.Constants
import com.mcal.common.data.ReactivePreferences
import com.mcal.common.filesystem.FilePickHelper
import com.mcal.common.utils.*
import com.mcal.common.utils.ScopedStorage.getProjects
import com.mcal.common.view.ProgressDialog.ProcessingInterface
import com.mcal.downloader.DownloaderActivity
import com.mcal.webview.WebViewActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import kotlin.system.exitProcess

class MainActivity : CustomizedLangActivity(), ProcessingInterface {
    private var _binding: ActivityMainBinding? = null
    private val binding get() = _binding!!
    private lateinit var pickLauncher: ActivityResultLauncher<Intent>

    companion object {
        private const val TAG = "MainActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupToolbar(id = R.id.toolbar, title = getString(R.string.app_name), back = false)
        addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_main, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                when (menuItem.itemId) {
                    R.id.action_settings -> {
                        startActivity(Intent(this@MainActivity, SettingsActivity::class.java))
                        return true
                    }
                    R.id.action_night_mode -> {
                        lifecycleScope.launch {
                            if (ReactivePreferences.isLegacyNightMode()) {
                                ReactivePreferences.setNightMode(false)
                                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                                delegate.applyDayNight()
                            } else {
                                ReactivePreferences.setNightMode(true)
                                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                                delegate.applyDayNight()
                            }
                        }
                        return true
                    }
                }
                return false
            }
        })

        initUI()

        if (BuildConfig.SHOW_AGREEMENT) {
            // Политика конфиденциальности
            if (!appLicenseAccepted(this)) {
                AppAgreementDialog(this)
            } else {
                initFileWithPermissionCheck()
            }
        } else {
            initFileWithPermissionCheck()
        }

        // Выбор АПК файла для полного редактирования
        pickLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                result.data?.data?.let { uri ->
                    val projectsDir = getProjects()
                    val pickedName = FilePickHelper.getFileName(this, uri) ?: "coping.apk"
                    val tmpApk = File(projectsDir, ".coping_$pickedName")
                    try {
                        contentResolver.openInputStream(uri)?.use { inputStream ->
                            copyFile(inputStream, tmpApk)
                        } ?: return@let

                        // Reject truncated / non-zip copies instead of crashing later
                        if (!isValidApk(tmpApk)) {
                            Toast.makeText(this@MainActivity, R.string.msg_unsupported_file, Toast.LENGTH_SHORT).show()
                            return@let
                        }

                        var name = "app"
                        // Получение имени приложения
                        ApkInfoParser().parse(this@MainActivity, tmpApk.path)?.label?.let {
                            name = it
                        }
                        val apk = File(projectsDir, "$name.apk")
                        // Копирование АПК во временное хранилище
                        if (!tmpApk.renameTo(apk)) {
                            copyFile(tmpApk, apk)
                        }
                        // Диалог с выбором режима декомпиляции
                        selectFullEditDialog(this@MainActivity, apk.path)
                    } catch (e: Exception) {
                        e.printStackTrace()
                        Toast.makeText(this@MainActivity, R.string.msg_unsupported_file, Toast.LENGTH_SHORT).show()
                    } finally {
                        if (tmpApk.exists()) {
                            tmpApk.delete()
                        }
                    }
                }
            }
        }

        scheduleCleaning()
    }

    public override fun onDestroy() {
        _binding = null
        super.onDestroy()
    }

    private fun initUI() {
        findViewById<MaterialToolbar>(R.id.toolbar).subtitle = Utils.getVersionString()

        binding.btnApk.setOnClickListener { pickApk() }
        binding.btnApp.setOnClickListener {
            startActivity(Intent(this, UserAppActivity::class.java))
        }
        binding.btnPrj.setOnClickListener {
            startActivity(Intent(this, ProjectListActivity::class.java))
        }
        binding.btnOdex.setOnClickListener {
            startActivity(Intent(this, OdexPatchActivity::class.java))
        }
        binding.btnTools.setOnClickListener {
            startActivity(Intent(this, DownloaderActivity::class.java))
        }
        binding.btnInfo.setOnClickListener {
            val intent = Intent(this, WebViewActivity::class.java)
            intent.putExtra("htmlUrl", Constants.getDomain() + "/apkeditor/doc/instructions/index.html")
            startActivity(intent)
        }
        binding.btnSettings.setOnClickListener {
            startActivity(Intent(this, SettingsActivity::class.java))
        }
        binding.btnExit.setOnClickListener {
            Process.killProcess(Process.myPid())
            exitProcess(0)
        }

        if (!ScopedStorage.isToolsInstalled()) {
            showToolManagerDialog()
        }
    }

    private fun pickApk() {
        pickLauncher.launch(FilePickHelper.pickFile(true))
    }

    private fun isValidApk(file: File): Boolean {
        if (!file.exists() || file.length() < 4L) return false
        return try {
            file.inputStream().use { input ->
                val magic = ByteArray(4)
                var offset = 0
                while (offset < magic.size) {
                    val read = input.read(magic, offset, magic.size - offset)
                    if (read < 0) return false
                    offset += read
                }
                magic[0] == 0x50.toByte() && magic[1] == 0x4B.toByte() &&
                    ((magic[2] == 0x03.toByte() && magic[3] == 0x04.toByte()) ||
                        (magic[2] == 0x05.toByte() && magic[3] == 0x06.toByte()) ||
                        (magic[2] == 0x07.toByte() && magic[3] == 0x08.toByte()))
            }
        } catch (e: Exception) {
            false
        }
    }

    private fun showToolManagerDialog() {
        val context = this@MainActivity
        val params = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )

        val padding = App.dp2px(16f, context).toInt()
        val container = LinearLayout(context).apply {
            layoutParams = params
            orientation = LinearLayout.VERTICAL
            setPadding(padding, padding, padding, 0)
        }

        MaterialAlertDialogBuilder(context)
            .setTitle(getString(R.string.dialog_install_frameworks))
            .setMessage(getString(R.string.dialog_install_frameworks_sum))
            .setView(container)
            .setCancelable(false)
            .setPositiveButton(android.R.string.ok) { dialog, _ ->
                val intent = Intent(context, DownloaderActivity::class.java)
                startActivity(intent)
                dialog.dismiss()
            }.show()
    }

    fun initFileWithPermissionCheck() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            != PackageManager.PERMISSION_GRANTED
        ) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                ActivityCompat.requestPermissions(
                    this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.POST_NOTIFICATIONS), 1
                )
            }
        }
    }

    // Очистка мусора исходя из заданного лимита
    private fun scheduleCleaning() {
        CoroutineScope(Dispatchers.IO).launch {
            // Общий размер всех папок в мегабайтах
            val total = Utils.getFoldersSize(
                ScopedStorage.cacheDir, ScopedStorage.getBackupsDir(),
                ScopedStorage.getDecodedDir(), ScopedStorage.getTmpDir()
            ) / 1000 / 1000
            Log.i(TAG, "scheduleCleaning: cache size = $total")
            if (total > ReactivePreferences.getGarbageLimit()) {
                Utils.deleteFiles(
                    ScopedStorage.cacheDir, ScopedStorage.getBackupsDir(),
                    ScopedStorage.getDecodedDir(), ScopedStorage.getTmpDir()
                )
            }
        }
    }

    @Throws(Exception::class)
    override fun process() {
        try {
            val intent = Intent(this, ApkComposeService::class.java)
            stopService(intent)
            val fileDir = filesDir
            val rootDirectory = fileDir.absolutePath
            val decodeRootPath = "$rootDirectory/decoded"
            deleteAll(File(decodeRootPath))
        } catch (e: Throwable) {
            e.printStackTrace()
        }
    }

    override fun afterProcess() {
        Process.killProcess(Process.myPid())
    }
}